jpa/eclipselink.jpa.test/src/org/eclipse/persistence/testing/tests/jpa/beanvalidation/BeanValidationJunitTest.java

```diff
@@ -342,6 +343,32 @@ public void testTraversableResolverPreventsLoadingOfLazyRelationships() {
         assertTrue( "Lazy field should not be instantiated because of validation", !isInstantiated(employee, "projects", project) );
         assertTrue( "Lazy field should not be instantiated because of validation", !isInstantiated(employee, "managedProject", project) );
     }
+    
+    public void testPesistWithInvalidAndModifiedBeanData() {
```

Comment:
Is this a different test than `BeanValidationJunitTest.testUpdateWithInvalidData()`? They seem to be trying to test the same thing, but `testUpdateWithInvalidData` doesn't seem to be using callbacks and is instead expecting validation on transaction commit. Is this right?

https://github.com/eclipse-ee4j/eclipselink/blob/b87d11af0889564f8da540bec3cdf98c14cd9894/jpa/eclipselink.jpa.test/src/org/eclipse/persistence/testing/tests/jpa/beanvalidation/BeanValidationJunitTest.java#L252-L258

According to 3.6 of the spec:

    These lifecycle validation events occur immediately after the point at which all the 
    PrePersist, PreUpdate, and PreRemove lifecycle callback method invocations respectively 
    have been complete

I'm wondering if this means `testUpdateWithInvalidData` is wrong.

Reply:
I saw a javadoc somewhere saying that current behaviour is expected and intended by the spec - prepersist is called on the first persist call and not during transaction commit, so subsequent updates to the entity, is is done here are not validated as some people would expect. I just don't know, where I saw that comment :-/

Reply:
here it is: https://github.com/eclipse-ee4j/eclipselink/blob/master/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/listeners/BeanValidationListener.java#L74

Reply:
@lukasj I see! I think there is a misconception then that validation occurs on transaction commit. After reading the spec, it seems pretty clear that validation occurs on the EntityManager operation invocations. However, I searched around online for some examples and it seems transaction commit is being expected by users. Given the comment you found, it certainly sounds like the current spec design has issues that we may want to address in the future.

For example, here is a [Hibernate example](https://www.thoughts-on-java.org/automatically-validate-entities-with-hibernate-validator/) I found. It seems the example they are using is expecting transaction commit. Perhaps Hibernate is not adhering to the spec? Idk.